ring/func/base.py

```diff
@@ -91,27 +91,31 @@ def coerce_function(t):
     # but NEVER add a general iterator processing. it will cause user bugs.
 
 
-def coerce(v):
+def coerce(v, inmemory_storage):
     """Transform the given value to cache-friendly string data."""
+
     type_coerce = coerce_function(type(v))
     if type_coerce:
         return type_coerce(v)
 
     if hasattr(v, '__ring_key__'):
         return v.__ring_key__()
 
+    if inmemory_storage and type(v).__hash__ != object.__hash__:
+        return hash(v)
+
     cls = v.__class__
     if cls.__str__ != object.__str__:
         return str(v)
 
     raise TypeError(
         "The given value '{}' of type '{}' is not a key-compatible type. "
-        "Add __ring_key__() or __str__().".format(v, cls))
+        "Add __ring_key__() or __str__() or __hash__().".format(v, cls))
```

Comment:
`__hash__()` only works for in-memory storage. Error messages need to consider it.

A or B or C -> A, B or C.

To keep the message generating easy, keep the original message and append an additional sentence for in-memory storages.